snap/validate_test.go

```diff
@@ -132,6 +154,149 @@ func (s *ValidateSuite) TestValidateAppName(c *C) {
+	app := createSampleApp()
+	delete(app.Plugs, "network-bind")
+	err := ValidateApp(app)
+	c.Assert(
+		err, ErrorMatches,
+		`"network-bind" interface plug is required when sockets are used`)
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsEmptyListenStream(c *C) {
+	app := createSampleApp()
+	app.Sockets["sock"].ListenStream = ""
+	err := ValidateApp(app)
+	c.Assert(err, ErrorMatches, `socket "sock" must define "listen-stream"`)
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsInvalidName(c *C) {
+	app := createSampleApp()
+	app.Sockets["sock"].Name = "invalid name"
+	err := ValidateApp(app)
+	c.Assert(err, ErrorMatches, `invalid socket name: "invalid name"`)
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsValidListenStreamAddresses(c *C) {
+	app := createSampleApp()
+	validListenAddresses := []string{
+		// socket paths using variables as prefix
+		"$SNAP_DATA/my.socket",
+		"$SNAP_COMMON/my.socket",
+		// abstract sockets
+		"@snap.mysnap.my.socket",
+		// addresses and ports
+		"8080",
+		"127.0.0.1:8080",
+		"[::]:8080",
+		"[::1]:8080",
+	}
+	socket := app.Sockets["sock"]
+	for _, validAddress := range validListenAddresses {
+		socket.ListenStream = validAddress
+		err := ValidateApp(app)
+		c.Check(err, IsNil, Commentf(validAddress))
+	}
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsInvalidListenStreamPath(c *C) {
+	app := createSampleApp()
+	invalidListenAddresses := []string{
+		// socket paths out of the snap dirs
+		"/some/path/my.socket",
+		"/var/snap/mysnap/20/my.socket", // path is correct but has hardcoded prefix
+	}
+	socket := app.Sockets["sock"]
+	for _, invalidAddress := range invalidListenAddresses {
+		socket.ListenStream = invalidAddress
+		err := ValidateApp(app)
+		c.Assert(err, ErrorMatches, `socket "sock" has invalid "listen-stream": only.*are allowed`)
+	}
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsInvalidListenStreamPathContainsDots(c *C) {
+	app := createSampleApp()
+	app.Sockets["sock"].ListenStream = "$SNAP/../some.path"
+	err := ValidateApp(app)
+	c.Assert(
+		err, ErrorMatches,
+		`socket "sock" has invalid "listen-stream": "\$SNAP/../some.path" should be written as "some.path"`)
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsInvalidListenStreamPathRelative(c *C) {
+	app := createSampleApp()
+	invalidListenAddresses := []string{
+		"$SNAP/my.socket", // snap dir is not writable
+		"$SOMEVAR/my.socket",
+	}
+	socket := app.Sockets["sock"]
+	for _, invalidAddress := range invalidListenAddresses {
+		socket.ListenStream = invalidAddress
+		err := ValidateApp(app)
+		c.Assert(
+			err, ErrorMatches,
+			`socket "sock" has invalid "listen-stream": only \$SNAP_DATA and \$SNAP_COMMON prefixes are allowed`)
+	}
+}
+
+func (s *ValidateSuite) TestValidateAppSocketsInvalidListenStreamAbstractSocket(c *C) {
+	app := createSampleApp()
+	invalidListenAddresses := []string{
+		"@snap.notmysnap.my.socket",
+		"@some.other.name",
+		"@snap.myappiswrong.foo",
```

Comment:
We should also test `@snap.myapp`, `@@snap.myapp.foo` and `@snap.myapp\0.foo`

Reply:
Done